edb/pgsql/resolver/command.py

```diff
@@ -339,6 +369,409 @@ def _preprocess_insert_stmt(
+    assert isinstance(sub_name, sn.QualName)
+    sub_source_name = sub_source.get_name(ctx.schema)
+    assert isinstance(sub_source_name, sn.QualName)
+    sub_target_name = sub_target.get_name(ctx.schema)
+    assert isinstance(sub_target_name, sn.QualName)
+
+    sub_name = link.get_shortname(ctx.schema)
+
+    ql_sub_source_ref = qlast.ObjectRef(
+        name=sub_source_name.name,
+        module=sub_source_name.module,
+    )
+    ql_sub_target_ref = qlast.ObjectRef(
+        name=sub_target_name.name,
+        module=sub_target_name.module,
+    )
+
+    ql_stmt: qlast.Expr = qlast.UpdateQuery(
+        subject=qlast.Path(steps=[ql_sub_source_ref]),
+        where=qlast.BinOp(  # ObjectType == value.source
+            left=qlast.Path(steps=[ql_sub_source_ref]),
+            op='=',
+            right=qlast.Path(steps=[value_ql]),
+        ),
+        shape=[
+            qlast.ShapeElement(
+                expr=qlast.Path(steps=[qlast.Ptr(name=sub_name.name)]),
+                operation=(
+                    qlast.ShapeOperation(op=qlast.ShapeOp.APPEND)
+                    if (
+                        link.get_cardinality(ctx.schema)
+                        == qltypes.SchemaCardinality.Many
+                    )
+                    else qlast.ShapeOperation(op=qlast.ShapeOp.ASSIGN)
+                ),
+                compexpr=qlast.Shape(
+                    expr=qlast.TypeCast(
+                        expr=qlast.Path(  # target
+                            steps=[value_ql, qlast.Ptr(name=sub_name.name)]
+                        ),
+                        type=qlast.TypeName(maintype=ql_sub_target_ref),
+                    ),
+                    elements=[
+                        qlast.ShapeElement(
+                            expr=qlast.Path(
+                                steps=[
+                                    qlast.Ptr(name=ptr_name, type='property')
+                                ],
+                            ),
+                            compexpr=qlast.Path(
+                                steps=[
+                                    value_ql,
+                                    qlast.Ptr(name=sub_name.name),
+                                    qlast.Ptr(name=ptr_name, type='property'),
+                                ],
+                            ),
+                        )
+                        for ptr_name, _ in value_columns
+                        if ptr_name not in ('source', 'target')
+                    ],
+                ),
+            )
+        ],
+    )
+    if not is_value_single:
+        # value relation might contain multiple rows
+        # to express this in EdgeQL, we must wrap `insert` into a `for` query
+        ql_stmt = qlast.ForQuery(
+            iterator=qlast.Path(steps=[qlast.IRAnchor(name=value_name)]),
+            iterator_alias=iterator_name,
+            result=ql_stmt,
+        )
+    # ql_stmt = qlast.Path(steps=[ql_stmt, qlast.Ptr(name=sub_name.name)])
```

Comment:
Since we are generating an UPDATE, can we do it without the iterator?
It might still be easiest to do it with the iterator, though.

Reply:
I don't think so.

The value relation might contains multiple different `source` values, which cannot be translated to a non-iterated update.